### PR TITLE
MDEV-27969 - buf_page_cleaner thread is being signaled even if

### PR DESCRIPTION
**Pull request by mysqlonarm**
- [x] *The Jira issue number for this PR is: MDEV-27969*

## Description
MDEV-27969 - buf_page_cleaner thread is being signaled even if dirty_pct > innodb_max_dirty_pages_pct_lwm

- buf_page_cleaner is signaled when the pages are added to the flush list.
  (starting 10.8.3 as part of mtr_t::commit)

- as part of the MDEV-27774 changes, the logic to signal buf_page_cleaner
  accidentally got missed due to pass-by-value issue
  (expected pass-by-reference).

  [mtr_t::commit flow would signal page cleaner if there are modified
   blocks. The object that captured modified blocks was expected to get
   passed by reference but accidentally got passed by value and so
   page-cleaner thread was never signaled from the calling block].

## How can this PR be tested?
- execute any write workload and monitor of pages are flushed when dirty_pct > innodb_max_dirty_pages_pct_lwm

## Basing the PR against the correct MariaDB version
- [ ] *This is a new feature and the PR is based against the latest MariaDB development branch*
- [X ] *This is a bug fix and the PR is based against the earliest branch in which the bug can be reproduced*

## Backward compatibility
N.A.

### CODE REVIEW

File: storage/innobase/mtr/mtr0mtr.cc [hunk 1]
```diff
@@ -306,6 +304,7 @@ struct DebugCheck {
 		return(true);
 	}
 };
+DebugCheck dummyDebugCheck;
```
Comment (dr-m) [inline storage/innobase/mtr/mtr0mtr.cc:307]: This could be `static`, right?

File: storage/innobase/mtr/mtr0mtr.cc [hunk 2]
```diff
@@ -468,8 +468,10 @@ void mtr_t::commit()
       mysql_mutex_unlock(&buf_pool.flush_list_mutex);
     }
   }
-  else
-    m_memo.for_each_block_in_reverse(CIterate<ReleaseAll>());
+  else {
+    ReleaseAll rl;
+    m_memo.for_each_block_in_reverse(CIterate<ReleaseAll>(rl));
+  }
```
Comment (dr-m) [inline storage/innobase/mtr/mtr0mtr.cc:474]: Monty wants us to deviate from the K&R formatting style by adding a line break before each `{` (except in the `switch` statement).

**Review by dr-m — COMMENTED** [review]
Thank you, this is a very good catch.

In the commit message, you could directly mention
```
commit a635c40648519fd6c3729c9657872a16a0a20821 (MDEV-27774)
```
The GitHub web interface would then create a hyperlink to the previous commit from the commit message, which is more convenient, in case someone wants to browse the history via that.

File: storage/innobase/mtr/mtr0mtr.cc (hunk 1 above)
Comment (mysqlonarm) [reply storage/innobase/mtr/mtr0mtr.cc:307]: done

File: storage/innobase/mtr/mtr0mtr.cc (hunk 2 above)
Comment (mysqlonarm) [reply storage/innobase/mtr/mtr0mtr.cc]: done

**Comment by mysqlonarm** [comment]
> If I understood it correctly, the problem is that `ReleaseBlocks` was being passed by value, and as a result, the check `rb.modified` would never hold in `mtr_t::commit()`. This would be good to mention explicitly in the commit message. After all, the bulk of the changes are related to the removal of the pass-by-value code.
> 
> I will try to see if I can fix that bug while allowing pass-by-value for the remaining uses of `CIterate`. To my understanding, passing small objects by value can be more efficient (and allow more optimizations at compilation time) than passing them by reference.

**Comment by dr-m** [comment]
Thank you for pointing out the problem. I pushed a simpler fix as c1e88a635dcacc094a02b6f6ca399cc2a3cca542. The main reason why `CIterate` exists is that unlike `Iterate`, it enables the stateless iteration without having to declare local variables.

This bug shows that we really should set up some code coverage monitoring.

**Pull request closed without merging**